src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs

```diff
@@ -3226,48 +3189,135 @@ private static Node BalanceNode(Node node)
                 }
                 else
                 {
-                    if (left != null)
-                    {
-                        _left = left;
-                    }
+                    _left = left;
+                    _right = right;
+                    _height = ParentHeight(left, right);
+                    _count = ParentCount(left, right);
+                    return this;
+                }
+            }
 
-                    if (right != null)
-                    {
-                        _right = right;
-                    }
+            /// <summary>
+            /// Creates a node mutation, either by mutating this node (if not yet frozen) or by creating a clone of this node
+            /// with the described changes.
+            /// </summary>
+            /// <param name="left">The left branch of the mutated node.</param>
+            /// <returns>The mutated (or created) node.</returns>
+            private Node MutateLeft(Node left)
+            {
+                Requires.NotNull(left, nameof(left));
+                Debug.Assert(!this.IsEmpty);
 
-                    _height = checked((byte)(1 + Math.Max(_left._height, _right._height)));
-                    _count = 1 + _left._count + _right._count;
+                if (_frozen)
+                {
+                    return new Node(_key, left, _right);
+                }
+                else
+                {
+                    _left = left;
+                    _height = ParentHeight(left, _right);
+                    _count = ParentCount(left, _right);
+                    return this;
+                }
+            }
+
+            /// <summary>
+            /// Creates a node mutation, either by mutating this node (if not yet frozen) or by creating a clone of this node
+            /// with the described changes.
+            /// </summary>
+            /// <param name="right">The right branch of the mutated node.</param>
+            /// <returns>The mutated (or created) node.</returns>
+            private Node MutateRight(Node right)
+            {
+                Requires.NotNull(right, nameof(right));
+                Debug.Assert(!this.IsEmpty);
+
+                if (_frozen)
+                {
+                    return new Node(_key, _left, right);
+                }
+                else
+                {
+                    _right = right;
+                    _height = ParentHeight(_left, right);
+                    _count = ParentCount(_left, right);
                     return this;
                 }
             }
 
             /// <summary>
+            /// Calculates the height of the parent node from its children.
+            /// </summary>
+            /// <param name="left">The left child.</param>
+            /// <param name="right">The right child.</param>
+            /// <returns>The height of the parent node.</returns>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
@danmosemsft I did a little more investigation into this. It looks like even though the method is visually small, `checked` actually generates 3 branches here (2 for the adds, 1 for the byte cast):

```cs
    public class Program
    {
        public static void Main()
        {
            C.ParentHeight(new C(), new C());
        }
        
        public class C
        {
            private byte _height;

            [MethodImpl(MethodImplOptions.NoInlining)]
            public static byte ParentHeight(C left, C right) => checked((byte)(1 + left._height + right._height));
        }
    }
```

Here is the generated asm:

```asm
; Assembly listing for method C:ParentHeight(ref,ref):ubyte
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T00] (  3,   3  )     ref  ->  rcx
;  V01 arg1         [V01,T01] (  3,   3  )     ref  ->  rdx
;  V02 OutArgs      [V02    ] (  1,   1  )  lclBlk (32) [rsp+0x00]
;
; Lcl frame size = 40

G_M29434_IG01:
       4883EC28             sub      rsp, 40

G_M29434_IG02:
       0FB64108             movzx    rax, byte  ptr [rcx+8]
       83C001               add      eax, 1
       7014                 jo       SHORT G_M29434_IG04
       0FB65208             movzx    rdx, byte  ptr [rdx+8]
       03C2                 add      eax, edx
       700C                 jo       SHORT G_M29434_IG04
       A900FFFFFF           test     eax, -256
       7505                 jne      SHORT G_M29434_IG04

G_M29434_IG03:
       4883C428             add      rsp, 40
       C3                   ret

G_M29434_IG04:
       E8B2F1AB5F           call     CORINFO_HELP_OVERFLOW
       CC                   int3

; Total bytes of code 39, prolog size 4 for method C:ParentHeight(ref,ref):ubyte
; ============================================================
```

In total the method comes out to be 39 bytes. So it may not be unreasonable for the JIT to not inline this after all, given that it has no way of knowing this is going to be used in a hot codepath.